src/Altinn.Apps/AppTemplates/AspNet/App.IntegrationTests/ApiTests/DataApiTest.cs

```diff
@@ -870,6 +873,71 @@ public async Task Data_Put_DataValuesUpdated_ExistingValueOverwritten()
+        /// but it states that it should be possible to combine the two.
+        /// </summary>
+        [Fact]
+        public async Task Instance_CustomDataValuesAdded_CustomValueMergedIn()
+        {
+            // Arrange
+            string org = "ttd";
+            string app = "datafields-app";
+            string instanceGuid = "447ed22d-67a8-42c7-8add-cc35eba304f1";
+            string dataGuid = "590ebc27-246e-4a0a-aea3-4296cb231d78";
+            string token = PrincipalUtil.GetToken(1337);
+
+            int expectedDataValuesCountAfterAutomaticUpdate = 2;
+            int expectedDataValuesCountAfterManualUpdate = 3;
+            string expectedKey = "AnotherField";
+            string expectedValue = "160694";
+            string expectedCustomKey = "customKey";
+            string expectedCustomValue = "customValue";
+
+            TestDataUtil.PrepareInstance(org, app, 1337, new Guid(instanceGuid));
+
+            HttpClient client = SetupUtil.GetTestClient(_factory, org, app);
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
+
+            // Act - automatic update of data values done as a result of configuration and triggered by a change in the data.
+            string dataRequestUri = $"/{org}/{app}/instances/1337/{instanceGuid}/data/{dataGuid}?dataType=default";
+            string dataRequestBody = "{\"skjemanummer\":\"1472\",\"spesifikasjonsnummer\":\"9812\",\"blankettnummer\":\"AFP-01\",\"tittel\":\"ArbeidsgiverskjemaAFP\",\"gruppeid\":\"8818\",\"OpplysningerOmArbeidstakerengrp8819\":{\"Arbeidsforholdgrp8856\":{\"AnsattSammenhengendeAnsattAnsettelsedatadef33267\":{\"value\":\"SophieSalt\",\"orid\":\"33267\"},},\"Skjemainstansgrp8854\":{\"Journalnummerdatadef33316\":{\"value\":\"160694\"}}}}";
+            HttpRequestMessage dataHttpRequestMessage = new HttpRequestMessage(HttpMethod.Put, dataRequestUri) { Content = new StringContent(dataRequestBody, Encoding.UTF8, "application/json") };
+
+            await client.SendAsync(dataHttpRequestMessage);
+
+            HttpResponseMessage responseAfterDataUpdate = await client.GetAsync($"/{org}/{app}/instances/1337/{instanceGuid}");
+            string contentAfterDataUpdate = await responseAfterDataUpdate.Content.ReadAsStringAsync();
+            Instance instanceAfterDataUpdate = JsonConvert.DeserializeObject<Instance>(contentAfterDataUpdate);
+
+            // Assert - after automatic update
+            Assert.Equal(HttpStatusCode.OK, responseAfterDataUpdate.StatusCode);
+            Assert.Equal(expectedDataValuesCountAfterAutomaticUpdate, instanceAfterDataUpdate.DataValues.Count);
+            Assert.True(instanceAfterDataUpdate.DataValues.ContainsKey(expectedKey));
+            Assert.Equal(expectedValue, instanceAfterDataUpdate.DataValues[expectedKey]);
+
+            // Act - manually updating data values with custom values by calling the service.
```

Comment:
Seems like the this second section tests the mock rather than the logic used in the apps? Do we need this?

Reply:
Yes, hence the comment above the test: 

"The test is not ideal as it actually tests the ServiceMock and not the real implementation, but it states that it should be possible to combine the two."

So totally agree, testing automatic configuration with manuell adding of data values would require integration tests where the platform isn't mocked.